reyna/src/com/b2msolutions/reyna/Repository.java

```diff
@@ -31,36 +33,92 @@ public void onCreate(SQLiteDatabase db) {
 
 	@Override
 	public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
-        Logger.v(TAG, "onUpgrade");
+		Logger.v(TAG, "onUpgrade");
 	}
 
 	public void insert(Message message) {
-        Logger.v(TAG, "insert");
-		if (message == null)
+		Logger.v(TAG, "insert");
+		if (message == null) {
+			return;
+		}
+
+		SQLiteDatabase db = this.getWritableDatabase();
+		this.insertMessage(db, message);
+	}
+
+	public void insert(Message message, long dbSizeLimit) {
+		Logger.v(TAG, "insert");
+		if (message == null) {
 			return;
+		}
+
+		SQLiteDatabase db = this.getWritableDatabase();
+		long dbSize = this.getDbSize(db);
+
+		//todo handle (dbSize > dbSizeLimit) with a shrink db operation instead
+		//of doing just one deletion
+		if (dbSize > dbSizeLimit || Math.abs(dbSizeLimit - dbSize) < (dbSize * 0.2)) { //0.2 = 20% of the db size
+			this.clearOldRecords(db, message);
+		}
+
+		this.insertMessage(db, message);
+	}
+
+	private long getDbSize(SQLiteDatabase db) {
+		Cursor cursor = db.rawQuery("pragma page_count", null);
+		cursor.moveToNext();
+		long pageCount = cursor.getLong(0);
+		cursor.close();
+
+		return pageCount * db.getPageSize();
+	}
 
-		SQLiteDatabase db = null;
+	private void insertMessage(SQLiteDatabase db, Message message) {
 		try {
-			db = this.getWritableDatabase();
 			db.beginTransaction();
 			ContentValues values = new ContentValues();
 			values.put("url", message.getUrl());
 			values.put("body", message.getBody());
 
-			long messageid = db.insert("Message", null, values);			
-			this.addHeaders(db, messageid, message.getHeaders());
+			long messageId = db.insert("Message", null, values);
+			this.addHeaders(db, messageId, message.getHeaders());
 			db.setTransactionSuccessful();
 
-            Logger.i("reyna", "Repository: inserted message " + messageid);
-		} finally {
-			if (db != null) {
+			Logger.i("reyna", "Repository: inserted message " + messageId);
+		}
+		finally {
+			if (db != null && db.inTransaction()) {
 				db.endTransaction();
 			}
 		}
 	}
 
+	private void clearOldRecords(SQLiteDatabase db, Message message) {
```

Comment:
can findOldestMessageIdWithType and clearOldRecords combined in one SQL query, to save disk read and CPU.

@Youhana-Hana